colour/volume/macadam_limits.py

```diff
@@ -142,3 +143,197 @@ def is_within_macadam_limits(
+        middle_opti_colour = 235
+        opti_colour[
+            middle_opti_colour
+            - half_width : middle_opti_colour
+            + half_width
+            + 1
+        ] = 1
+        # we start the construction of the optimum color
+        # at the center of the opti_colour-array
+        opti_colour = np.roll(
+            opti_colour, center_opti_colour - middle_opti_colour
+        )
+        # the optimum colour is rolled to the right wavelength
+        return opti_colour
+
+    def bright_opti_colour(width, center, lightsource):
+        brightness = (
+            np.sum(optimum_colour(width, center) * lightsource)
+            / maximum_brightness
+        )
+        return brightness
+
+    step_size = np.array([64, 32, 16, 8, 4, 2, 1])
+    for wavelength in range(0, 471):
+        width = 127
+        for n in step_size:
+            brightness = bright_opti_colour(width, wavelength, Y_illuminated)
+            if brightness > target_bright or width > 234:
+                width -= n
+            else:
+                width += n
+
+        brightness = bright_opti_colour(width, wavelength, Y_illuminated)
+        if brightness < target_bright:
+            width += 1
+
+        rough_optimum = optimum_colour(width, wavelength)
+        brightness = np.sum(rough_optimum * Y_illuminated) / maximum_brightness
+
+        # in the following, the both borders of the found rough_optimum
+        # are reduced to get more exact results
+        bright_difference = (brightness - target_bright) * maximum_brightness
+        # discrimination for single-wavelength-spectra
+        if width > 0:
+            opti_colour = np.zeros(471)
+            opti_colour[
+                middle_opti_colour - width : middle_opti_colour + width + 1
+            ] = 1
+            # instead rolling forward opti_colour, light is rolled backward
+            rolled_light = np.roll(
+                Y_illuminated, middle_opti_colour - wavelength
+            )
+            opti_colour_light = opti_colour * rolled_light
+            left_opti = opti_colour_light[middle_opti_colour - width]
+            right_opti = opti_colour_light[middle_opti_colour + width]
+            interpolation = 1 - (bright_difference / (left_opti + right_opti))
+            opti_colour[middle_opti_colour - width] = interpolation
+            opti_colour[middle_opti_colour + width] = interpolation
+            # opti_colour is rolled to right position
+            final_optimum = np.roll(
+                opti_colour, wavelength - middle_opti_colour
+            )
+        else:
+            final_optimum = rough_optimum / brightness * target_bright
+
+        out_X = np.sum(final_optimum * X_illuminated) / maximum_brightness
+        out_Y = target_bright
+        out_Z = np.sum(final_optimum * Z_illuminated) / maximum_brightness
+        triple = np.array([out_X, out_Y, out_Z])
+        out_limits[wavelength] = triple
+    return out_limits
```

Comment:
Looking at this implementation, I'm left wondering if we should not simply output the XYZ values from this definition which does a similar work but using Trimesh and geometrical intersection of the *Rösch-MacAdam* colour solid with a plane: https://github.com/colour-science/colour/blob/a412b670448534987fc25d3a1a09cedc2a3ea5f2/colour/plotting/section.py#L479

https://colour.readthedocs.io/en/develop/_images/Plotting_Plot_Visible_Spectrum_Section.png

Reply:
I was thinking the same thing after looking at the spectrum.py code in #994